hw/xbox/mcpx_apu.c

```diff
@@ -403,25 +408,42 @@ static const MemoryRegionOps vp_ops = {
     .write = vp_write,
 };
 
-static void scratch_rw(hwaddr sge_base, unsigned int max_sge,
+static void scratch_rw(MCPXAPUState *d,
+                       hwaddr sge_base, unsigned int max_sge,
                        uint8_t *ptr, uint32_t addr, size_t len, bool dir)
 {
-    int i;
-    for (i = 0; i < len; i++) {
-        unsigned int entry = (addr + i) / TARGET_PAGE_SIZE;
-        assert(entry < max_sge);
+    unsigned int page_entry = addr / TARGET_PAGE_SIZE;
+    unsigned int offset_in_page = addr % TARGET_PAGE_SIZE;
+    unsigned int bytes_to_copy = TARGET_PAGE_SIZE - offset_in_page;
```

Comment:
We could replace these division ops with `page_entry = addr >> TARGET_PAGE_BITS` and `offset_in_page = addr & (TARGET_PAGE_BITS-1)`, though the compiler will probably optimize to this anyhow.

Reply:
Yeah, I assumed the compiler will optimize this, and this version is probably more readable.

*I assume you mean `offset_in_page = addr & (TARGET_PAGE_SIZE-1)`, which also depends on alignment of the page-size.*